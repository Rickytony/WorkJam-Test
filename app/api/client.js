import axios from "axios";

const client = axios.create({
  baseURL: "https://hacker-news.firebaseio.com/v0"
});

export { client };
