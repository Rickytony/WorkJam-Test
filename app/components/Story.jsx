import React from "react";
import PropTypes from "prop-types";
import { getComment } from "../api/client.js";
import { CommentList } from "./CommentList.jsx";

class Story extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      comments: [],
      showComments: false
    };

    this.getComments = this.getComments.bind(this);
    this.toggleComments = this.toggleComments.bind(this);
  }

  getComments(commentIds) {
    commentIds.forEach(comment => {
      return getComment(comment)
        .then(response => {
          this.setState({
            comments: [...this.state.comments, response.data]
          });
        })
        .catch(error => {
          throw new Error(error);
        });
    });
  }

  toggleComments(e) {
    e.preventDefault();
    this.setState(prevState => ({
      showComments: !prevState.showComments
    }));
  }

  componentDidMount() {
    if (!this.props.data.kids) return;
    this.getComments(this.props.data.kids.slice(0, 20));
  }

  render() {
    const { data } = this.props;
    return (
      <div>
        <a href={data.url} target="_blank">
          <h5 className="col">{data.title}</h5>
        </a>
        <h6 className="col">by {data.by}</h6>
        <CommentList
          comments={this.state.comments}
          showComments={this.state.showComments}
          handleClick={this.toggleComments}
        />
      </div>
    );
  }
}

Story.propTypes = {
  data: PropTypes.shape({
    title: PropTypes.string.isRequired,
    by: PropTypes.string.isRequired
  })
};

export { Story };
