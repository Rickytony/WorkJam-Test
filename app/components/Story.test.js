import React from "react";
import { shallow } from "enzyme";
import * as client from "../api/client";
import { Story } from "./Story";

jest.mock("../api/client");

const story = { title: "story title", by: "author", id: 1, kids: [11, 22, 33] };
const comment = { by: "commentor", text: "foo bar", id: 1 };

client.getComment = jest.fn().mockResolvedValue(comment);

describe("Story", () => {
  it("renders correctly", () => {
    const wrapper = shallow(<Story story={story} />);
    expect(wrapper).toMatchSnapshot();
  });

  it("updates showComments when toggleComments is called", () => {
    const wrapper = shallow(<Story story={story} />);
    wrapper.setState({ showComments: false });
    wrapper.instance().toggleComments({ preventDefault() {} });
    expect(wrapper.state().showComments).toBe(true);
  });

  it("calls getComment", () => {
    const wrapper = shallow(<Story story={story} />);
    expect(client.getComment).toHaveBeenCalledWith(story.kids[0]);
  });
});
