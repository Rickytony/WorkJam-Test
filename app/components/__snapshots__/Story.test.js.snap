// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Story renders correctly 1`] = `ShallowWrapper {}`;
